generator/src/lib.rs

```diff
@@ -2373,33 +2371,25 @@ fn derive_getters_and_setters(
 
             #[doc(hidden)]
             #[deprecated = "Migrate to `push()` if `next` does not have an existing chain (i.e. `p_next` is `NULL`), `extend()` otherwise"]
-            pub unsafe fn push_next<T: #extends_name + ?Sized>(self, next: &'a mut T) -> Self {
+            pub unsafe fn push_next<T: StructExtends<Self> + ?Sized>(self, next: &'a mut T) -> Self {
                 self.extend(next)
             }
         }
     } else {
         quote!()
     };
 
-    // Root structs come with their own trait that structs that extend
-    // this struct will implement
-    let next_trait = if root_struct_next_field.is_some() {
-        quote!(pub unsafe trait #extends_name {})
-    } else {
-        quote!()
-    };
-
     let lifetime = has_lifetimes.contains(&name).then(|| quote!(<'a>));
 
     // If the struct extends something we need to implement the traits.
     let impl_extend_trait = struct_
         .extends
         .iter()
         .flat_map(|extends| extends.split(','))
-        .map(|extends| format_ident!("Extends{}", name_to_tokens(extends)))
         .map(|extends| {
+            let base = name_to_tokens(extends);
             // Extension structs always have a pNext, and therefore always have a lifetime.
-            quote!(unsafe impl #extends for #name<'_> {})
+            quote!(unsafe impl StructExtends<#base<'_>> for #name<'_> {})
```

Comment:
Does coupling these two lifetimes have any weird effects? I'm not sure if that's a change in the semantics of `push` impls or not. I guess the examples build fine...

Reply:
Good catch. I think there should be _some relation_, such that anything already borrowed by `#name<'a>` gets forwarded into the borrow of `#base<'b>` which I assume is what we end up getting in the `push` functions for `T: Extends<Self>`?

I came up with this little test to validate that:

```rust
let mut layers = vec![];
let mut api =
    vk::PhysicalDeviceLayeredApiPropertiesListKHR::default().layered_apis(&mut layers);
let pdev_props = vk::PhysicalDeviceProperties2::default().push(&mut api);

// Access to either variable is disallowed because both are mutably borrowed in pdev_props
dbg!(&api);
dbg!(&layers);

dbg!(pdev_props); // Holds a borrow on api and layers
```

Specifically `layers` should be inaccessible as long as `pdev_props` is live. This appears to be satisfied (by failed compilation) for each of these:

```rust
unsafe impl Extends<PhysicalDeviceProperties2<'_>> for PhysicalDeviceLayeredApiPropertiesListKHR<'_> {}
unsafe impl<'root, 'child> Extends<PhysicalDeviceProperties2<'root>> for PhysicalDeviceLayeredApiPropertiesListKHR<'child> {}
unsafe impl<'a> Extends<PhysicalDeviceProperties2<'a>> for PhysicalDeviceLayeredApiPropertiesListKHR<'a> {}
```

And I'm not exactly sure what is most correct. Perhaps the second case might lack a `'child: 'root`, i.e. the scope of `'child` is at least as large, or bigger than `'root` (`'child` can live longer than `'root` but not vice-versa. Likewise, as long as `'root` is `mut`ably borrowed, `'child` is `mut`ably borrowed as well).

:shrug: as usual, I could use a little bit of help/validation on lifetimes :)

Reply:
The `'_` version is equivalent to the single-variable `'a` version. I think the two-variable version is directly analogous to what we had before; the missing constraint is funny looking, but I believe exactly that is implied whenever you actually call `push` because `&'root Child<'child>` is illegal unless `'child: 'root`, so everything works out.

I think it does make sense that the single-variable version works too: we can always coerce `Child<'child>` to `Extension<'root>` if `'child: 'root`, so `fn push(&mut self, child: &'root Child<'root>)` is well-typed and has no adverse effects.

So, both approaches are at least superficially valid. I lean towards the two-variable approach:
- This trait describes a structural relationship between types, not a dynamic one, so saying things about lifetimes doesn't seem relevant
- There's no benefit in introducing a constraint that the signature of `push` already enforces
- It minimizes the change wrt. the previous semantics

Not a strongly held opinion, though. It's not exactly unintuitive to say that an extension definitionally outlives the thing it extends.

Reply:
We should have a `trybuild` test to validate that short-lived extension structs with long-lived roots are rejected.

Reply:
I added `trybuild`. But whenever having this snippet in `ash/src/lib.rs`:

```rust
#[test]
fn trybuild() {
    let t = trybuild::TestCases::new();
    // t.compile_fail("tests/fail/long_lived_root_struct_borrow.rs");
}
```

```console
❯ cargo t
   Compiling ash-window v0.13.0 (/newdata/Code/TraverseResearch/ash/ash-window)
   Compiling ash v0.38.0+1.3.296 (/newdata/Code/TraverseResearch/ash/ash)
   Compiling ash-examples v0.1.0 (/newdata/Code/TraverseResearch/ash/ash-examples)
error[E0282]: type annotations needed
   --> ash/src/extensions/khr/acceleration_structure.rs:266:77
    |
266 |         assert_eq!(max_primitive_counts.len(), build_info.geometry_count as _);
    |                                                                             ^ cannot infer type

For more information about this error, try `rustc --explain E0282`.
error: could not compile `ash` (lib test) due to 1 previous error
```

@Ralith any clue?

Reply:
Also, I love this crate and suggestion already. As mentioned I immediately moved the `push()` functions to `trait AnyTaggedStructure`, and had a `// TODO: update lifetimes` because the `fn push<T: Extends<Self> + ?Sized>(mut self, next: &mut T) -> Self` no longer captures the borrow of `T` in the returned `Self`, and the test caught it :+1:

Reply:
> Obviously the fix is easy - I just wish to understand why this wasn't a problem in the past. Does this crate affect assert_eq!()?

That is pretty weird. Maybe it's getting a very old rustc from your environment somewhere, or the build configuration is otherwise changing to add/remove some `PartialEq` impls in ways that introduce ambiguity? Not sure.

Reply:
It's not your environment, repros here too... must be a change in the visible impls somehow.

Reply:
And it occurs while compiling `ash (lib test)` before `trybuild` is executed. Unless `trybuild` uses some magic to inject itself whenever it's compiled/linked into our crate, but that should happen with it being present in `Cargo.toml` regardless.

I don't see any difference in `cargo t -v -j1` leading up to the error that stands out.